### ISSUE TITLE
Publish ports to ingress network after docker swarm init --force-new-cluster

### ISSUE THREAD

**Issue by destebanm** [moby/moby#34987]
We are using 17.06.2

I have discovered one issue when doing `docker swarm init --force-new-cluster`

We are running 3 managers node swarm in EC2, with a global service. This service publishes the 8080 port to the ingress network with `--publish 8080:8080.` It works fine. If I run  `iptables -t nat -L -n | grep 8080` in whichever node I get:
`DNAT tcp -- 0.0.0.0/0 0.0.0.0/0 tcp dpt:8080 to:172.18.0.2:8080
`
If we remove two nodes at the same time, the service continues working on only one node, but it works.

We run `docker node ls` in the healthy node we get The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online. This is expected. So the only solution is to run a `docker swarm init --force-new-cluster --listen-addr $PRIVATE_IP:2377 --advertise-addr $PRIVATE_IP:2377`

Now is when the issue arises: it seems that service starts without problems but the thing is that it does not publish the port to the ingress network. We checked by doing` iptables -t nat -L -n | grep 8080` and not getting any result.

This causes that our service is not available to listen in 8080

Any help to avoid that?

Thanks!

**Comment by thaJeztah** [comment]
Can you provide more information as requested in the [issue-template](https://raw.githubusercontent.com/docker/docker/master/.github/ISSUE_TEMPLATE.md) that was shown when you opened this issue.

- provide the output of `docker version`
- provide the output of `docker info`

Also relevant system and/or daemon logs from manager and worker nodes could be helpful (if possible in debug mode)

**Comment by destebanm** [comment]
I have updated the issue using the template

Also, I have done the tests using the new released 17.09

**Comment by destebanm** [comment]
Hey guys! Have you seen anything? Is there something you need from our side?

Thanks!

**Comment by thaJeztah** [comment]
ping @fcrisciani @abhi PTAL

**Comment by gshively11** [comment]
@thaJeztah @fcrisciani @abhi I can corroborate this bug and it is a showstopper for us. We're running on `17.09.1-ce, build 19e2cf6`. Due to some other unknown bug, our managers flipped out when doing a normal demote->drain->down rebuild and we couldn't recover. After trying `--force-new-cluster`, all our services were unreachable, even though they showed as online and healthy. I then reproduced the bug using the steps @destebanm listed above.

`--force-new-cluster` is the only option I'm aware of for an unrecoverable quorum that should result in minimal downtime. This seems like a critical requirement for swarm and I would love to see this bug fixed. I'm happy to try to contribute to the effort, although I'm not familiar with go. Any chance this can be tackled soon?

**Comment by gshively11** [comment]
I forgot to mention, after deleting all the services and re-deploying them, they're still unreachable.

**Comment by fcrisciani** [comment]
cc @ctelfer

**Comment by fcrisciani** [comment]
@gshively11 we found a regression starting from 17.09, @ctelfer is working on it, and we will link the PR as it is ready, would be also great if you will be able to test it, just to give confidence that the correct behavior is restored

**Comment by gshively11** [comment]
@fcrisciani excellent news! I will definitely be able to test it.

**Comment by gshively11** [comment]
A few more things:

None of our services are relying on the swarm load balancer.  Instead, we're using traefik (load balancer), binding that to host ports via the bridge, then having traefik proxy traffic to our services using their container ips/ports. They're all attached to a custom overlay network.

Before doing `--force-new-cluster`, I'm able to verify connections succeed between the traefik container and service containers using telnet, and that they fail after `--force-new-cluster`.

I took a peek at what I think is the [PR you mentioned?](https://github.com/docker/libnetwork/pull/2101) and I'm not so sure that's related to the bug I'm seeing (although as mentioned, I'm very unfamiliar with the codebase).

**Comment by gshively11** [comment]
I looked through the networks on the master and worker (bridge, our ingresses, and docker_gwbridge), as well as the nat iptable, before and after running `--force-new-cluster`, and the only differences I noticed were:
- A service on the master restarted with a new task id (don't think this is relevant)
- The peers reported on the worker's ingress networks no longer mentioned the master afterward:
before
```
"Peers": [
    {
        "Name": "master-name-redacted-01634da704d3",
        "IP": "10.38.12.3"
    },
    {
        "Name": "worker-name-redacted-757359ba62ac",
        "IP": "10.38.10.18"
    }
]
```
after
```
"Peers": [
    {
        "Name": "worker-name-redacted-757359ba62ac",
        "IP": "10.38.10.18"
    }
]
```

Everything else appeared the same on the networks and nat iptable. The master node missing from the peer list does seem concerning.

**Comment by fcrisciani** [comment]
@gshively11 once you force a new cluster you will need to join back all the nodes, it's a new cluster at this point

**Comment by gshively11** [comment]
I thought the point of `--force-new-cluster` was to force one manager to become the leader of a new cluster using the old clusters config, but also have all the other workers and such still joined to the new cluster with services running?  If I run `docker node ls` on the master after `--force-new-cluster`, I can still see the worker in the node list and I see all the services still running, which makes it seem like the existing worker is already part of the new cluster.

**Comment by gshively11** [comment]
Obviously the managers would need to be rejoined to recreate a new highly available quorum, the documentation says as much, but that shouldn't apply to workers.

**Comment by fcrisciani** [comment]
@gshively11 sorry for the confusion, I wanted to mean all the managers. When you say the master node is missing, which master were you talking about?

**Comment by fcrisciani** [comment]
ok now it's clear, it sounds a different signature, but maybe the 2 things are still related. @ctelfer once the patch is ready, can you verify if the operation of `--force-new-cluster` actually keeps the ingress properly configured?

**Comment by gshively11** [comment]
to clarify, none of our containers are actually connect to the ingress network. we have a user-defined attachable overlay network that all the containers are attached to, except for our traefik container, which publishes ports on the host in global mode. traefik then routes traffic to the container ips/ports via the user-defined overlay.

**Comment by gshively11** [comment]
more information. I've tried the following to get the traefik container on the manager talking to the service container on the worker after a `--force-new-cluster`, with no luck:

- drain and then active the worker
- drain, leave the swarm, then rejoin the swarm on the worker
- delete and re-deploy the service
- delete and re-deploy traefik
- delete both stacks (including the overlay network) and re-deploy

All of these attempts resulted in the services running, but the traffic between them still failing. It seems like `--force-new-cluster` just breaks overlay networks completely on the swarm.

**Comment by fcrisciani** [comment]
@gshively11 from the last 2 messages, this does not seem related to ingress, can you open a separate issue to track your case? this ticket is mentioning explicitly ingress network

**Comment by ctelfer** [comment]
FWIW, I have tested using a build based on https://github.com/moby/moby/pull/36538 and the fix did not prevent the ingress NAT mappings from vanishing.  As @fcrisciani indicated and as @gshively11 predicted, this seems to be a different issue.

**Comment by shibug** [comment]
I tried to recreate the firewall rules for DOCKER-INGRESS chain by comparing with one of the worker nodes but it is still not working. 

Is there a workaround to resolve this.

**Comment by shibug** [comment]
I am having the same issue while upgrading nodes from 17.12 to 18.03. Managers nodes were upgraded to 18.03 although I had to use force new cluster. Worker nodes are still at 17.12. 

3 manager and 6 worker nodes.

**Comment by shibug** [comment]
The workaround was to make the worker nodes leave the swarm and join again.  That started opening up the published ports on the ingress network.  Had to do a couple of nodes for all ports to publish because not all services were assigned to only one worker but all of them were assigned to those 2 workers that might have done the trick

**Comment by nbaum** [comment]
I've encountered this problem, and whilst I'm not really sure of the cause and effect, it was resolved immediately after flushing iptables and restarting docker.

```sh
systemctl stop docker # or however you do it
iptables -F # remove all the rules
iptables -X # remove all custom chains
systemctl start docker # or however you do it
```

**Comment by mrmachine** [comment]
This is still an issue, and yes it is a show-stopper.

I'm running Docker for AWS and I changed instance type for managers from m4.large to t2.small. That change was completed successfully by adding a new manager then removing the old, one at a time, always keeping at least 3 managers running.

5 minutes after that finished, the auto scaling group terminated one of the new managers because `an instance was taken out of service in response to a ELB system health check failure.` 

Possibly due to some other unknown bugs, the replacement manager node was unable to re-join the swarm and all 3 manager nodes were both reporting `The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online.`

Perhaps the cause of the failure was that by temporarily increasing the manager count to 4, then decreasing it to 3 the swarm was already running in a "degraded" state (missing one?) and then when the auto scaling group removed another (down to 2) we no longer had more than half running. But anyway, that's not relevant to the main issue.

Once quorum was lost, services on worker nodes were still accessible via an HAproxy service on one of the manager nodes.

I ran `docker swarm init --force-new-cluster` and as others have observed, HAproxy (and all other services) on that manager node was restarted, but was no longer accessible from outside. I could no longer connect to HAproxy via the Amazon ELB that Docker for AWS creates as part of its stack, nor directly via the manager node's IP address.

I tried updating the HAproxy service/stack, which worked but did not restore external connectivity.

I tried removing and recreating the HAproxy stack, which worked but did not restore external connectivity.

I tried to have the the other 2 managers leave their swarm to rejoin the new swarm, but they just froze on every attempt to join the new swarm.

In the end, I had to reboot (via AWS console) every manager node, force leave the swarm, create a new swarm from scratch, rejoin the other two manager nodes, recreate all my stacks and services from scratch, and force leave the worker nodes and rejoin them to the new swarm as well.

All this resulted in 1.5 hours downtime on our production cluster, and it's lucky I only had a handful of stacks to recreate. On another cluster we have many more stacks and I would not like to be recreating them all from scratch when recovering from a similar failure.

If `--force-new-cluster` had worked as expected, there could have been literally zero downtime in recovering from this failure.

Here's the version info:

```
~ $ docker version
Client:
 Version:	18.03.0-ce
 API version:	1.37
 Go version:	go1.9.4
 Git commit:	0520e24
 Built:	Wed Mar 21 23:05:35 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.03.0-ce
  API version:	1.37 (minimum version 1.12)
  Go version:	go1.9.4
  Git commit:	0520e24
  Built:	Wed Mar 21 23:14:32 2018
  OS/Arch:	linux/amd64
  Experimental:	true
~ $ docker info
Containers: 9
 Running: 7
 Paused: 0
 Stopped: 2
Images: 18
Server Version: 18.03.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: kqhbm8wu7776yeane23x24uj4
 Is Manager: true
 ClusterID: rneqw02gox7x0gzmgbn4p73ig
 Managers: 3
 Nodes: 5
 Orchestration:
  Task History Retention Limit: 2
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 172.31.22.183
 Manager Addresses:
  172.31.22.183:2377
  172.31.33.167:2377
  172.31.8.171:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfd04396dc68220d1cecbe686a6cc3aa5ce3667c
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.81-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.95GiB
Name: ip-172-31-22-183.us-west-2.compute.internal
ID: VMKC:YLGI:5SMC:LK3C:FX6A:LGND:CRM2:7CSQ:FSXA:TWQG:5WJH:2F3H
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 instance_type=t2.small
 node_type=manager
 os=linux
 region=us-west-2
 availability_zone=us-west-2b
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

~ $ 
```

**Comment by jandsu** [comment]
I believe I have the same issue on my day-1 using Swarm mode...

Question is how to recover from 3 manager nodes when two are down? To simplify the description, I create the swarm with 2 nodes and take one down.

Here are the steps to reproduce on 2 VMs on my dev box, in the hope that this may help reproduce and debug the issue.

## Create the swarm

From host1 (10.33.16.135):

    # docker swarm init
    # docker swarm join-token manager

From host2 (10.33.16.159):

    # docker swarm join --token SWMTKN-1-32wze40bpzepvnnr9gt0vfi5zgllhss16uaosfed53014x9t35-5h6s5npwggvwzu9l2ih4u13wk 10.33.16.135:2377

From host2:
```
# docker node ls
ID                            HOSTNAME  STATUS  AVAILABILITY  MANAGER STATUS  ENGINE VERSION
yhrxg3hbpyqpdgszz96bf0q14     host1     Ready   Active        Leader          18.06.1-ce
yzxrie9ntewc765vfuog0on6a *   host2     Ready   Active        Reachable       18.06.1-ce
```

then

```
# docker stack deploy --compose-file docker-compose.yaml foo
```

where docker-compose.yaml is simply:
```
version: '3.7'

networks:
  my-net:
    driver: overlay

services:
  nginx:
    image: nginx
    ports:
      - 80:80
    networks:
      - my-net
    deploy:
      replicas: 1
```

```
# docker stack ps foo
ID            NAME         IMAGE          NODE     DESIRED STATE       CURRENT STATE            ERROR               PORTS
pxkg8msqqbpn  foo_nginx.1  nginx:latest   host2    Running             Running 38 seconds ago
```

## Check that it works

```
# curl -I http://host1
HTTP/1.1 200 OK
Server: nginx/1.15.3
Date: Mon, 17 Sep 2018 15:12:08 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Tue, 28 Aug 2018 13:32:13 GMT
Connection: keep-alive
ETag: "5b854edd-264"
Accept-Ranges: bytes
```

```
# curl -I http://host2
HTTP/1.1 200 OK
Server: nginx/1.15.3
Date: Mon, 17 Sep 2018 15:12:10 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Tue, 28 Aug 2018 13:32:13 GMT
Connection: keep-alive
ETag: "5b854edd-264"
Accept-Ranges: bytes
```

## Now forcibly power off host1 to break the raft consensus

Observe that the service on the still running node is still running.

```
# curl -I http://host2
HTTP/1.1 200 OK
Server: nginx/1.15.3
Date: Mon, 17 Sep 2018 15:15:20 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Tue, 28 Aug 2018 13:32:13 GMT
Connection: keep-alive
ETag: "5b854edd-264"
Accept-Ranges: bytes
```

but... (as expected)
```
# docker node ls
Error response from daemon: rpc error: code = Unknown desc = The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online.
```

Now, you **have to** apply the recipe for the lost raft consensus.

Indeed, if you start another host (say host3) and try to join it, you get the same error (__the swarm does not have a leader__) so (from host2)
```
# docker swarm init --force-new-cluster

# docker node ls
ID                            HOSTNAME  STATUS  AVAILABILITY  MANAGER STATUS
yhrxg3hbpyqpdgszz96bf0q14     host1     Down    Active                      
yzxrie9ntewc765vfuog0on6a *   host2     Ready   Active        Leader        

# docker stack ps foo
ID                  NAME                IMAGE               NODE     DESIRED STATE       CURRENT STATE            ERROR               PORTS
lvo7wkg8jk9a        foo_nginx.1         nginx:latest        host2    Running             Running 2 minutes ago                        
pxkg8msqqbpn         \_ foo_nginx.1     nginx:latest        host2    Shutdown            Complete 2 minutes ago               
```

which looks good :) ...

... but the service is broken :(

```
# curl -I http://host2
curl: (7) Failed to connect to host2 port 80: Connection refused
```
------
## Docker version
```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:51 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:23:15 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```
## Docker info
```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: yzxrie9ntewc765vfuog0on6a
 Is Manager: true
 ClusterID: 1o1zq5onla87gplygv4xv6c1y
 Managers: 2
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 10.33.16.159
 Manager Addresses:
  10.33.16.135:2377
  10.33.16.159:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-34-generic
Operating System: Ubuntu 18.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.853GiB
Name: host2
ID: PATK:EVZJ:H4D5:2FRB:B4C4:JR66:QVOC:PEDE:HQMD:RAFM:5KZR:IR4V
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support

```

**Comment by dchiaramello** [comment]
Hello.

Are there any news about that problem?

I'm facing the same issue with:
Docker version 18.09.7, build 2d0083d

Regards,
Daniel

**Comment by rchen9012** [comment]
I've encountered the same problem today, after issue "--force-new-cluster" then make all worker leave swarm and 
rejoin the swarm as worker.  deploy new stack, stack publish to default ingree network, ip subnet is same as previous mounted stacks .but no access to the new stack.  any solution?  docker versions 19.03.12, build 48a66213fe
